### PR TITLE
Design proposal: ambient cache initialization speed improvement

### PR DESCRIPTION
**Pull request by geolives**
Added design proposal for ambient cache initialization speed improvement to fix issue https://github.com/maplibre/maplibre-native/issues/1815

I recreated a pull request to isolate the design proposal for the implementation. Sorry for the inconvenience.
Original pull request here : https://github.com/maplibre/maplibre-native/pull/1839

### CODE REVIEW

**Comment by louwers** [comment]
@JesseCrocker Was [your comment](https://github.com/maplibre/maplibre-native/pull/1839#issuecomment-1797043203) answered to your satisfaction by [this answer](https://github.com/maplibre/maplibre-native/pull/1839#issuecomment-1798864147)?

File: design-proposals/2023-11-06-ambient-cache-initialization-speed-improvement.md [hunk 1]
```diff
@@ -0,0 +1,53 @@
+# Design Proposal : Ambient cache initialization speed improvement
+
+## Motivation
+
+We would like to improve the speed of ambient cache initialization when there is a lot of downloaded offline regions.
```
Comment (louwers) [inline design-proposals/2023-11-06-ambient-cache-initialization-speed-improvement.md:5]:
```suggestion
We would like to improve the speed of ambient cache initialization when a lot of offline regions are downloaded.
```

File: design-proposals/2023-11-06-ambient-cache-initialization-speed-improvement.md [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+The problem is described in this issue : https://github.com/maplibre/maplibre-native/issues/1815
```
Comment (louwers) [inline design-proposals/2023-11-06-ambient-cache-initialization-speed-improvement.md:7]:
```suggestion
The problem is described in [this issue](https://github.com/maplibre/maplibre-native/issues/1815).
```

File: design-proposals/2023-11-06-ambient-cache-initialization-speed-improvement.md [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+## Proposed Change
+
+We propose to split the ambient cache from the offline database cache.
+In terms of database structure, that means create two new tables. Let's say ambient_tiles and ambient_resources.
```
Comment (louwers) [inline design-proposals/2023-11-06-ambient-cache-initialization-speed-improvement.md:12]:
```suggestion
In terms of database structure, that means two new tables are created. They will be called `ambient_tiles` and `ambient_resources`.
```

File: design-proposals/2023-11-06-ambient-cache-initialization-speed-improvement.md [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+### When we need to store resource for ambient cache:
+
+We store it into the ambient_resources table.
```
Comment (louwers) [inline design-proposals/2023-11-06-ambient-cache-initialization-speed-improvement.md:16]:
```suggestion
We store it into the `ambient_resources` table.
```

File: design-proposals/2023-11-06-ambient-cache-initialization-speed-improvement.md [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+### When we need to store tile for ambient cache:
+
+We store it into the ambient_tiles table.
```
Comment (louwers) [inline design-proposals/2023-11-06-ambient-cache-initialization-speed-improvement.md:20]:
```suggestion
We store it into the `ambient_tiles` table.
```

File: design-proposals/2023-11-06-ambient-cache-initialization-speed-improvement.md [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+- We would clear the ambient cache at first upgrade to the new database version => Some slowness may occur only on the first initialization after the upgrade
+
+## Rejected Alternatives
+
+2 possible alternatives to solve this issue:
+
+- Store the current ambient cache size inside the database, and update it each time his cache size is modified. This way, the initialization simply request the computed cache size from the last execution.
+- Store a flag inside the tiles and resources table that tells if the resource/tile is ambient or not. That way we simplify the request by deleting the need to make a left join. With the cost that for each download/cleaning, the flag must be re-set correctly.
+
+The main problem we have with these 2 solutions are in term of database corruption. If something wrong happens between the initial modification and the computation modification, you'll have a database with incoherent information stored. That could leads to issues difficult to overcome.
```
Comment (louwers) [inline design-proposals/2023-11-06-ambient-cache-initialization-speed-improvement.md:53]:
```suggestion
The main problem we have with these 2 solutions are related to database corruption. If something wrong happens between the initial modification and the computation modification, you'll have a database with incoherent information stored. That could lead to issues difficult to overcome.
```

File: design-proposals/2023-11-06-ambient-cache-initialization-speed-improvement.md [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+We store it into the ambient_tiles table.
+
+### When we need to store resource for offline data:
+
+We store it into the resources table with the associated region inside the region_resources table.
+We delete the same resource from the ambient cache (*).
+
+### When we need to store tile for offline data:
+
+We store it into the tile table with the associated region inside the region_tiles table.
```
Comment (louwers) [inline design-proposals/2023-11-06-ambient-cache-initialization-speed-improvement.md:29]:
```suggestion
We store it into the tile table with the associated region inside the `region_tiles` table.
```

File: design-proposals/2023-11-06-ambient-cache-initialization-speed-improvement.md [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+We store it into the ambient_resources table.
+
+### When we need to store tile for ambient cache:
+
+We store it into the ambient_tiles table.
+
+### When we need to store resource for offline data:
+
+We store it into the resources table with the associated region inside the region_resources table.
```
Comment (louwers) [inline design-proposals/2023-11-06-ambient-cache-initialization-speed-improvement.md:24]:
```suggestion
We store it into the resources table with the associated region inside the `region_resources` table.
```

**Review by louwers — APPROVED** [review]
Looks OK to me, made some textual suggestions.

**Comment by geolives** [comment]
Hello.

I have an additional question about the implementation of the changes in the deletion of offline regions:

With our changes, when an offline region is deleted, tiles and resources which are not used anymore in any other offline region are not moved to the ambient cache again (which would impact the performance because we would have to move the data from tiles / resources tables to ambient_tiles / ambient_resources) but simply removed from the tiles / resources tables in the database.

However, it causes a behaviour change : these resources and tiles are deleted from the database, they don't return to the ambient cache, where the deletion should take the LRU mechanism into account.

Is this behaviour change acceptable?

If it is not acceptable, what would you suggest to avoid impacting the performance?

Thanks in advance.

**Comment by louwers** [comment]
When deleting a locally stored region for offline usage, I think it is expected that that region needs to be re-downloaded. I wonder if @JesseCrocker agrees?

**Pull request merged**